Support big creation flag

tonysun83 opened the pull request:
Erlang 23 and above requires support for DFLAG_BIG_CREATION.
The node understands big node creation tags NEW_PID_EXT, NEW_PORT_EXT
and NEWER_REFERENCE_EXT. So that means we need to be able to decode
and encode those new formats.

tonysun83 commented:
Need to run testy tests between erl 23 dbcore and clouseau

tonysun83 commented:
Ran Classic with erl 23.3.1:
```
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [hipe]
```

```
suites/search/test_lucene_queries.py::TestLuceneQueries::test_bookmark_empty PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_bookmark_has_no_rows PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_bookmark_has_rows PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_bookmark_invalid PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_counts PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_drilldown PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_group_field_number_is_unsupported PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_group_field_string PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_group_field_string_sort_ascending PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_group_field_string_sort_descending PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_group_field_string_with_group_limit PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_group_field_string_with_row_limit PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_include_doc_false PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_include_fields PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_include_fields_invalid_fields PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_include_x_doc PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_limit_of_0 PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_limit_of_x PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_query PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_ranges PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_result_json_structure PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_sort_ascending_number PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_sort_ascending_string PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_sort_descending_number PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_sort_descending_string PASSED
suites/search/test_lucene_queries.py::TestLuceneQueries::test_stale_ok PASSED
```

rnewson commented on src/main/scala/scalang/epmd/EpmdDecoder.scala:
```diff
@@ -25,10 +25,14 @@ class EpmdDecoder extends FrameDecoder {
         if (buffer.readableBytes < 4) return null
         val result = buffer.getByte(1)
-        val creation = buffer.getUnsignedShort(2)
+        val creation = if (header == 118) {
```
I think it's more readable to handle each case separately, even if there's a bit of duplication.

rnewson commented on src/main/scala/scalang/node/ScalaTermEncoder.scala:
```diff
@@ -89,6 +89,14 @@ class ScalaTermEncoder(peer: Symbol, encoder: TypeEncoder = NoneTypeEncoder) ext
     case s : Symbol =>
       writeAtom(buffer, s)
+    case Reference(node, id, creation : Int) => //we only emit new references
+      buffer.writeByte(90)
```
we should only send the new kind of reference if the connection has the flag set

tonysun83 replied:
@rnewson I know that the distribution flags are agreed upon at handshake time. 
```
Early in the distribution handshake the two participating nodes exchange capability flags. This is done in order to determine how the communication between the two nodes should be performed. The intersection of the capabilities presented by the two nodes defines the capabilities that will be used. 
```

But I'm not sure where to retrieve that info in the context of this call:
```
encode(ctx : ChannelHandlerContext, channel : Channel, obj : Any)
```
I'll check tomorrow how JInterface does it.

rnewson replied:
in our situation it may not matter. the receiving node is erlang 20 and can handle the 'new' message types.

tonysun83 commented:
@rnewson I've removed the encodings and only used the new encodings.

I tested the newer  build with older erlang 20 dbcore and I saw no errors with a normal `testy` run. So that gives me confidence during our upgrades.

I added a `exit(boom)` during a search request here: https://github.com/cloudant-labs/clouseau/blob/master/src/main/scala/com/cloudant/clouseau/IndexService.scala#L352 to induce a monitor exit. On the clouseau side, we trap the exit and remove the process for the lru: https://github.com/cloudant-labs/clouseau/blob/master/src/main/scala/com/cloudant/clouseau/IndexManagerService.scala#L160
One thing to note is that in erl 20, the monitor message was still being received (because that one testcase passes). So I think the behavior will stay the same from Clouseau's perspective.

From the dreyfus_index side, the exit of the sibling clouseau process leads to an exit as expected:
https://github.com/apache/couchdb/blob/3.1.x/src/dreyfus/src/dreyfus_index.erl#L210

```
index for <<"shards/a0000000-bfffffff/adm/testy_db_ftvpkmapdg.1631854012 _design/views101 animals">> closed with reason boom
```


It also causes this `DOWN` clause to be exercised with a reason = `killed`, https://github.com/apache/couchdb/blob/3.1.x/src/dreyfus/src/dreyfus_index.erl#L232
```
[info] 2021-09-17T04:46:51.931913Z node1@127.0.0.1 <0.12091.0> -------- Shutting down index server <<"shards/a0000000-bfffffff/adm/testy_db_ypwrlryalz.1631854005 _design/views101 min_length">>, db <0.11700.0> closing w/ reason killed
```
It's a local process to this node, so it means it's a dreyfus process, but I'm curious what process it monitored.

I didn't see any dreyfus_index_manager mentions in the logs so I don't think monitor exits propagate back up to that level. 

The logs also showed rexit errors:

```
[error] 2021-09-17T04:46:45.609134Z node3@127.0.0.1 emulator -------- Error in process <0.11585.0> on node 'node3@127.0.0.1' with exit value:
{{error,{badmatch,{'EXIT',killed}}},[{mem3_rpc,rexi_call,3,[{file,"src/mem3_rpc.erl"},{line,392}]},{mem3_rep,update_locals,3,[{file,"src/mem3_rep.erl"},{line,609}]},{mem3_rep,replicate_batch,3,[{file,"src/mem3_rep.erl"},{line,523}]},{maps,map_1,2,[{file,"maps.erl"},{line,253}]},{maps,map,2,[{file,"maps.erl"},{line,244}]},{mem3_rep,replicate_batch_multi,1,[{file,"src/mem3_rep.erl"},{line,506}]},{mem3_rep,'-push_changes/1-fun-0-',3,[{file,"src/mem3_rep.erl"},{line,450}]},{mem3_rep,with_src_db,2,[{file,"src/mem3_rep.erl"},{line,715}]}]}

```
but I think these are expected if we are forcefully terminating the search processes.

tonysun83 commented:
@rnewson

> dreyfus a) receives a monitor exit message for an open index 

Our dreyfus_index processes are only linked to the clouseau process. So shouldn't we only get EXIT messages if clouseau crashes? 

> b) does the right thing when that happens, though your test probably covers quite a lot of the same ground.

I killed the jvm to fake a crash and read the logs. I saw `noconnection exit messages as usual but the logs don't have any new monitoring messages. Upon restarting the clouseau, things continued normally and indexing/search worked as normal.

rnewson commented:
Does killing the jvm cause it to send the monitor exit message you fixed?

tonysun83 commented:
@rnewson I don't think we can generate that encoded monitor message in any test with dreyfus/clouseau because that message is only gets sent to remote monitors. Dreyfus doesn't explicitly monitor clouseau process. Any clouseau process that dies without reason normal exit will ultimately be a local process that gets handled by trapMonitor:

https://github.com/cloudant/scalang/blob/cloudant2-version-bump/src/main/scala/scalang/Node.scala#L625-L628
->
https://github.com/cloudant/scalang/blob/cloudant2-version-bump/src/main/scala/scalang/Process.scala#L61-L63
->
https://github.com/cloudant-labs/clouseau/blob/master/src/main/scala/com/cloudant/clouseau/IndexManagerService.scala#L158

rnewson commented:
as noted elsewhere, dreyfus sends its requests to clouseau via the ioq queueing application. ioq monitors the target process before sending the message, and demonitors on receipt of the reply. It's interesting in the extreme that it turns out scalang was never capable of sending a valid monitor exit message in the entire time since cliff and I added monitoring support in 2012. The upside of this is that it suggests that other options, like JInterface, are not missing a valuable feature.

tonysun83 commented:
The `DOWN` message gets received here if we use ioq:

https://github.com/apache/couchdb-ioq/blob/main/src/ioq_server.erl#L157-L164

```
node1@127.0.0.1 <0.325.0> -------- DOWN MESSAGE RECEIVED REF: #Ref<0.1489396447.3362521095.222840>, REASON: boom
```

and when we use ioq2:

https://github.com/apache/couchdb-ioq/blob/main/src/ioq_server2.erl#L345-L352
```
IOQ2 DOWN MESSAGE RECEIVED REF: #Ref<0.442507440.3628859398.222783>, REASON: boom
```

In both instances, we're sending an EXIT message to the waiting processes on this IOQ call. 
In this case it's the dreyfus_index process which process it here:

https://github.com/apache/couchdb/blob/3.x/src/dreyfus/src/dreyfus_index.erl#L210-L218

This closes the index correctly.


On the clouseau side, the local process monitoring traps the monitor exit. Any remote monitors should get a message for the exit. To confirm both, I added additional logging.
```
clouseau.main [INFO] Trapping Monitor Exit for Reason 'boom

scalang.ErlangNode [INFO] sending remote handling monitor exit for Monitor(Pid('node1@127.0.0.1,325,0,1631807826),Pid('clouseau1@127.0.0.1,179,1,2),Reference('node1@127.0.0.1,Vector(137625, 1751384069, 389121099),1631807826)), reason 'boom
```

tonysun83 commented:
keeping all commits to reflect required changes and for transparency
